### PR TITLE
Update

### PR DESCRIPTION
**Pull request by OmerKahani**
Refactor the create application method so it will also do update and delete.

Not sure about the name of the function, does doing delete in an apply function is ok?

### CODE REVIEW

File: pkg/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -111,21 +145,30 @@ func (r *ApplicationSetReconciler) createApplications(ctx context.Context, appli
 		})
 
 		if err != nil {
-			log.Error(err, fmt.Sprintf("failed to get Application %s resource for applicationSet %s", app.Name, applicationSetInfo.Name))
+			log.Error(err, fmt.Sprintf("failed to CreateOrUpdate Application %s resource for applicationSet %s", app.Name, applicationSetInfo.Name))
```
Comment (xianlubird) [inline pkg/controllers/applicationset_controller.go:148]: We should add `action` in the log

File: pkg/controllers/applicationset_controller.go [hunk 2]
```diff
@@ -111,21 +145,30 @@ func (r *ApplicationSetReconciler) createApplications(ctx context.Context, appli
...
+		r.Recorder.Eventf(&applicationSetInfo, core.EventTypeNormal, fmt.Sprint(action), "%s Application %q", action, app.Name)
 		log.Infof("%s Application %s resource for applicationSet %s", action, app.Name, applicationSetInfo.Name)
 	}
 
-	return nil
-
-}
-
-func (r *ApplicationSetReconciler) getApplications(ctx context.Context, applicationSetInfo argoprojiov1alpha1.ApplicationSet) ([]argov1alpha1.Application, error) {
+	// Delete apps that are not in m[string]bool
```
Comment (xianlubird) [inline pkg/controllers/applicationset_controller.go:156]: I think we should separate out this function ,   delete application shouldn't exists in apply function

Comment (OmerKahani) [reply pkg/controllers/applicationset_controller.go:160]: done

File: pkg/controllers/applicationset_controller.go [hunk 3]
```diff
@@ -78,17 +79,34 @@ func (r *ApplicationSetReconciler) Reconcile(req ctrl.Request) (ctrl.Result, err
 
 	}
 
-	if err := r.createApplications(ctx, applicationSetInfo, desiredApplications); err != nil {
-		log.Infof("Unable to create applications applicationSetInfo %e", err)
-		return ctrl.Result{}, err
-	}
+	r.createOrUpdateInCluster(ctx, applicationSetInfo, desiredApplications)
+	r.deleteInCluster(ctx, applicationSetInfo, desiredApplications)
```
Comment (xianlubird) [inline pkg/controllers/applicationset_controller.go:83]: Should we have some `if conditions`, like if `finalizers` is set , we should delete this cluster?

Comment (dgoodwin) [reply pkg/controllers/applicationset_controller.go:83]: Not sure if this is what you meant, we definitely will want to add a finalizer to ApplicationSets, and if deleting the appset we must cleanup all Applications before removing that finalizer. Deletion cleanup should probably be a separate PR. I can try to find some time for this one this week.

File: pkg/controllers/applicationset_controller.go (hunk 1 above)
Comment (OmerKahani) [reply pkg/controllers/applicationset_controller.go]: done

File: pkg/controllers/applicationset_controller.go (hunk 3 above)
Comment (OmerKahani) [reply pkg/controllers/applicationset_controller.go:83]: this function deletes old applications from the cluster (deleted record from a list)

Deletion of ApplicationSets is implemented with owner reference so the Kubernetes GC will catch it and delete the applications

File: pkg/controllers/applicationset_controller.go [hunk 4]
```diff
@@ -99,34 +117,60 @@ func (r *ApplicationSetReconciler) SetupWithManager(mgr ctrl.Manager) error {
...
 		found := app
 		action, err := ctrl.CreateOrUpdate(ctx, r.Client, &found, func() error {
 			found.Spec = app.Spec
-			return controllerutil.SetControllerReference(&applicationSetInfo, &found, r.Scheme)
+			return controllerutil.SetControllerReference(&applicationSet, &found, r.Scheme)
 		})
 
 		if err != nil {
-			log.Error(err, fmt.Sprintf("failed to get Application %s resource for applicationSet %s", app.Name, applicationSetInfo.Name))
+			log.Error(err, fmt.Sprintf("failed to CreateOrUpdate Application %s resource for applicationSet %s", app.Name, applicationSet.Name))
```
[the remark this answers is not in the document]
Comment (OmerKahani) [reply pkg/controllers/applicationset_controller.go]: added thanks!

File: pkg/controllers/applicationset_controller.go (hunk 3 above)
Comment (OmerKahani) [reply pkg/controllers/applicationset_controller.go:83]: going over the proposal, every application should have `resources-finalizer.argocd.argoproj.io` entry in `metadata.finalizers` 

Do we also want to support the option to delete a record from a list / directory from git, and not run the finalizers in the application (something like `kubectl delete app NAME --cascade=false`) ?

Comment (xianlubird) [reply pkg/controllers/applicationset_controller.go:83]: > Do we also want to support the option to delete a record from a list / directory from git, and not run the finalizers in the application (something like kubectl delete app NAME --cascade=false) ?

I think we don't need to implement this for now

Comment (dgoodwin) [reply pkg/controllers/applicationset_controller.go:83]: > this function deletes old applications from the cluster (deleted record from a list)
> 
> Deletion of ApplicationSets is implemented with owner reference so the Kubernetes GC will catch it and delete the applications

My mistake, forgot about that!

**Comment by dgoodwin** [comment]
LGTM. I'll merge it in.

**Pull request merged**